compiler/rustc_mir_build/src/build/matches/mod.rs

```diff
@@ -1130,6 +1137,33 @@ impl<'tcx, 'pat> Candidate<'pat, 'tcx> {
             |_| {},
         );
     }
+
+    pub(crate) fn coverage_id(&self) -> coverage::CandidateCovId {
+        self.coverage_id
+    }
+
+    pub(crate) fn set_coverage_id(&mut self, coverage_id: coverage::CandidateCovId) {
+        self.coverage_id = coverage_id;
+        // Assign id for match pairs only if this candidate is the root.
+        if !coverage_id.subcandidate_id.is_root() {
+            return;
+        };
+        let mut match_pairs_count = 0;
+        let mut next_match_id = || {
+            match_pairs_count += 1;
+            coverage::MatchPairId(match_pairs_count)
+        };
+        let mut match_pairs = self.match_pairs.iter_mut().collect::<Vec<_>>();
+        while let Some(match_pair) = match_pairs.pop() {
+            match_pair.coverage_id = next_match_id();
```

Comment:
Is this just a unique identifier for the match pair? If so, I would prefer to set it at the creation of the match pair.

Reply:
Yeah, we only need this identifier to be unique in the match arm.

Reply:
Looks like `MatchPair` could be created manywhere. Do you advise to assign them the identifiers on creating candidates or just make the identifier global unique?

Reply:
`MatchPair::new` appears to be called in many places, but almost all of those calls are directly/indirectly recursive.

The only external caller of `MatchPair::new` is `FlatPat::new`.

Reply:
I think it would better not set identifiers of `MatchPair` untill the `coverage_id` of candidate is set. 
This can avoid overhead when mcdc is not enabled by options, or ignores some candidates (for instance, candidates from iteration). 
What do you think?

Reply:
I haven't looked at this closely yet, but I had a thought:

Is it possible to associate each MatchPair with the THIR id of its corresponding pattern? If so, that might avoid the need to have a separate set of IDs for coverage tracking.